File: iosApp/iosApp/data/mappers/channels/SupergroupMapper.swift
//
//  SupergroupMapper.swift
//  iosApp
//
//  Created by nVoxel on 13.05.2024.
//  Copyright © 2024 orgName. All rights reserved.
//

import Foundation
import TDLibKit
import TGDriveKit

class SupergroupMapper {
    
    private let photoMapper: PhotoMapper
    
    init(photoMapper: PhotoMapper) {
        self.photoMapper = photoMapper
    }
    
    func toResponse(supergroup: Supergroup) -> CommonSupergroup {
        var isCreator = false
        
        switch (supergroup.status) {
        case .chatMemberStatusCreator(let creatorInfo):
            isCreator = true
        default:
            isCreator = false
        }
        
        return CommonSupergroup(
            supergroupId: supergroup.id,
            username: supergroup.usernames?.editableUsername,
            date: Int32(supergroup.date),
            isCreator: isCreator,
            memberCount: Int32(supergroup.memberCount),
            isChannel: supergroup.isChannel
        )
    }
    
    func toResponse(supergroupFullInfo: SupergroupFullInfo) -> CommonSupergroupFullInfo {
        return CommonSupergroupFullInfo(
            chatPhoto: supergroupFullInfo.photo == nil ? nil : toResponse(chatPhoto: supergroupFullInfo.photo!),
            description: supergroupFullInfo.description,
            chatInviteLink: supergroupFullInfo.inviteLink == nil ? nil : toResponse(chatInviteLink: supergroupFullInfo.inviteLink!)
        )
    }
    
    private func toResponse(chatPhoto: ChatPhoto) -> CommonChatPhoto {
        return CommonChatPhoto(
            id: chatPhoto.id.rawValue,
            addedDate: Int32(chatPhoto.addedDate),
            size: photoMapper.toResponse(sizes: chatPhoto.sizes)
        )
    }
    
    private func toResponse(chatInviteLink: ChatInviteLink) -> CommonChatInviteLink {
        return CommonChatInviteLink(
            link: chatInviteLink.inviteLink,
            name: chatInviteLink.name,
            creatorUserId: chatInviteLink.creatorUserId,
            expirationDate: Int32(chatInviteLink.expirationDate),
            isPrimary: chatInviteLink.isPrimary,
            isRevoked: chatInviteLink.isRevoked
        )
    }
}

File: source/data/mappers/src/jvmMain/kotlin/com/voxeldev/tgdrive/mappers/channels/SupergroupMapper.kt
package com.voxeldev.tgdrive.mappers.channels

import com.voxeldev.tgdrive.mappers.media.PhotoMapper
import org.drinkless.tdlib.TdApi.ChatInviteLink
import org.drinkless.tdlib.TdApi.ChatMemberStatusCreator
import org.drinkless.tdlib.TdApi.ChatPhoto
import org.drinkless.tdlib.TdApi.Supergroup
import org.drinkless.tdlib.TdApi.SupergroupFullInfo
import com.voxeldev.tgdrive.files.api.channels.Supergroup as CommonSupergroup
import com.voxeldev.tgdrive.files.api.channels.SupergroupFullInfo as CommonSupergroupFullInfo
import com.voxeldev.tgdrive.main.api.chats.ChatInviteLink as CommonChatInviteLink
import com.voxeldev.tgdrive.main.api.chats.ChatPhoto as CommonChatPhoto

/**
 * @author nvoxel
 */
class SupergroupMapper(
    private val photoMapper: PhotoMapper,
) {

    fun toResponse(supergroup: Supergroup): CommonSupergroup =
        CommonSupergroup(
            supergroupId = supergroup.id,
            username = supergroup.usernames?.editableUsername,
            date = supergroup.date,
            isCreator = supergroup.status is ChatMemberStatusCreator,
            memberCount = supergroup.memberCount,
            isChannel = supergroup.isChannel,
        )

    fun toResponse(supergroupFullInfo: SupergroupFullInfo): CommonSupergroupFullInfo =
        CommonSupergroupFullInfo(
            chatPhoto = supergroupFullInfo.photo?.toResponse(),
            description = supergroupFullInfo.description,
            chatInviteLink = supergroupFullInfo.inviteLink?.toResponse(),
        )

    private fun ChatPhoto.toResponse(): CommonChatPhoto =
        CommonChatPhoto(
            id = id,
            addedDate = addedDate,
            size = photoMapper.toResponse(sizes = sizes),
        )

    private fun ChatInviteLink.toResponse(): CommonChatInviteLink =
        CommonChatInviteLink(
            link = inviteLink,
            name = name,
            creatorUserId = creatorUserId,
            expirationDate = expirationDate,
            isPrimary = isPrimary,
            isRevoked = isRevoked,
        )
}
